### ISSUE TITLE
Didn't show method or line number for %method [%line], instead got "? [?]"

### ISSUE THREAD

**Issue by danielwnn** [huorswords/Microsoft.Extensions.Logging.Log4Net.AspNetCore#38]
Is there any package missing?

```c#
  <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.1" />
  <PackageReference Include="log4net" Version="2.0.8" />
  <PackageReference Include="Microsoft.Extensions.Logging.Log4Net.AspNetCore" Version="2.2.1" />

log4net.config below:

<?xml version="1.0" encoding="utf-8" ?>
<log4net>
  <appender name="ConsoleAppender" type="log4net.Appender.ManagedColoredConsoleAppender">
    <mapping>
      <level value="ERROR" />
      <foreColor value="Red" />
    </mapping>
    <mapping>
      <level value="WARN" />
      <foreColor value="Yellow" />
    </mapping>
    <mapping>
      <level value="INFO" />
      <foreColor value="White" />
    </mapping>
    <mapping>
      <level value="DEBUG" />
      <foreColor value="Green" />
    </mapping>    
    <layout type="log4net.Layout.PatternLayout">
      <conversionPattern value="%date %5level %logger.%method [%line] - MESSAGE: %message%newline %exception" />
    </layout>
  </appender>
    
  <appender name="RollingLogFileAppender" type="log4net.Appender.RollingFileAppender">
    <file value="LogFiles/"/>
    <appendToFile value="true"/>
    <rollingStyle value="Date"/>
    <maxSizeRollBackups value="10"/>
    <maximumFileSize value="100MB"/>
    <datePattern value="yyyy-MM-dd'.log'" />
    <staticLogFileName value="false"/>
    <layout type="log4net.Layout.PatternLayout">
      <conversionPattern value="%date %5level %logger.%method [%line] - MESSAGE: %message%newline %exception" />
    </layout>
  </appender>
  <root>
    <level value="ALL" />
    <appender-ref ref="ConsoleAppender" />
    <appender-ref ref="RollingLogFileAppender" />
  </root>
</log4net>

Code in Startup.cs:

        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            loggerFactory.AddLog4Net();

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseCookiePolicy();

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=EmployeeProfile}/{action=Index}/{id?}");
            });
        }
```
Console output:

 2018-07-01 13:04:40,217  INFO Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.? [?] - MESSAGE: Executed action method ProfileMgrApp.Controllers.EmployeeProfileController.Index (ProfileMgrApp), returned result Microsoft.AspNetCore.Mvc.ViewResult in 332.8077ms.
 2018-07-01 13:04:40,275  INFO Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.? [?] - MESSAGE: Executing ViewResult, running view Index.
 2018-07-01 13:04:40,541  INFO Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.? [?] - MESSAGE: Executed ViewResult - view Index executed in 298.2617ms.
 2018-07-01 13:04:40,563  INFO Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.? [?] - MESSAGE: Executed action ProfileMgrApp.Controllers.EmployeeProfileController.Index (ProfileMgrApp) in 817.0772ms
 2018-07-01 13:04:40,589  INFO Microsoft.AspNetCore.Hosting.Internal.WebHost.? [?] - MESSAGE: Request finished in 1221.1667ms 200 text/html; charset=utf-8

**Comment by huorswords** [comment]
Hello @danielwnn ,

I have been searching about your issue, and seems that could be related to the warning notes from log4net library related to appenders:

[Log4Net PatterLayout](https://logging.apache.org/log4net/log4net-1.2.13/release/sdk/log4net.Layout.PatternLayout.html)

> _Note about caller location information.
The following patterns %type %file %line %method %location %class %C %F %L %l %M all generate caller location information. Location information uses the System.Diagnostics.StackTrace class to generate a call stack. The caller's information is then extracted from this stack._
> _CAUTIONï¿½ï¿½ï¿½
The System.Diagnostics.StackTrace class is not supported on the .NET Compact Framework 1.0 therefore caller location information is not available on that framework._
> _CAUTIONï¿½ï¿½ï¿½
The System.Diagnostics.StackTrace class has this to say about Release builds:_
> _"StackTrace information will be most informative with Debug build configurations. By default, Debug builds include debug symbols, while Release builds do not. The debug symbols contain most of the file, method name, line number, and column information used in constructing StackFrame and StackTrace objects. StackTrace might not report as many method calls as expected, due to code transformations that occur during optimization."_
> _This means that in a Release build the caller information may be incomplete or may not exist at all! Therefore caller location information cannot be relied upon in a Release build._

I have been testing that is not related to the Release build at all, and no, it is not related. Then I tested if can be related to .NET Core compatibility from log4net package and... bingo!

I have created a test project targeting 4.6.2 with your appender configuration and the test logs show me the line number and the method correctly. Doing the same with a .NET Core 1.1 / 2.0 test project, the logs show ? symbol.

So, as far Microsoft.Extensions.Logging.Log4Net.AspNetCore project is only intended to allow use log4net library into the Microsoft.Extensions.Logging system, and it is not related to the log4net appenders itself, I'm sorry but I cannot help you further on this issue... :(

**Comment by danielwnn** [comment]
Thank you for the insights! I will try it with a debug build.

**Comment by heapifyman** [comment]
I'm facing the same issue and I have a question about the explanation why this issue has been closed.
Does the problem exist in Log4Net.AspNetCore (this library) or in log4net itself, i.e. should I switch to NLog instead?
Or does the problem exist in dotnet core itself and `%method` or `%line` cannot be supported in any logging framework?

**Comment by huorswords** [comment]
Hello @heapifyman,

It is a behavior of Log4Net appenders, not from this repo, as far as I know.

I don't really know if the behavior is reproduced on NLog or Serilog providers, but I doubt it, because they are officially supported by the Microsoft official extensions.

Please, let me know if you test it on any of these logging libraries and the results.

Thank you.

**Comment by heapifyman** [comment]
I found this [issue](https://issues.apache.org/jira/browse/LOG4NET-606), almost one year old without any progress. So it seems to be a problem of log4net in combination with dotnet core 2.

I tried with NLog and [callsite-linenumber](https://github.com/NLog/NLog/wiki/Callsite-line-number-layout-renderer) and [callsite](https://github.com/NLog/NLog/wiki/Callsite-Layout-Renderer) work as expected. Class-/methodname and line number show up in the log entries.

Serilog does not seem to support linenumbers out of the box - but apparently it's possible to implement it yourself. Haven't tried it, though.

I think I'm gonna switch to NLog. It seems to be working fine, so far.
